### PR TITLE
Timer concurrency logic

### PR DESCRIPTION
**Pull request by roji**
As discussed, this redoes the timer management to (hopefully) remove all race conditions.

* The timer object is only allocated when starting replication, and disposed when exiting it.
* When disposing, we wait until callbacks have completed (this is a bit tricky to do asynchronously on netstandard2.0, but I think it's OK). This means that we can never have timer *callbacks* running when StartReplication has returned.
* A race condition may still make a feedback packet arrive at PG after replication has completed. I have done a check, and PG doesn't mind if a feedback message is received in idle state (outside of replication), so we should be good.
* The above also means that we don't need to suspend the timer when cancelling replication.
* Finally, it's worth noting that if SendFeedback gets stuck (because of a network partition), StartReplication will wait. As far as I can tell, this isn't a problem - we should already have timeouts in place for this case, and there's nothing special with this scenario compared to SendFeedback blocking anywhere else.

/cc @Brar and @vonzshik for an extra pair of eyes for this concurrency-sensitive business

### CODE REVIEW

File: src/Npgsql/Replication/ReplicationConnection.cs [hunk 1]
```diff
@@ -428,74 +421,100 @@ async Task<TimelineHistoryFile> TimelineHistoryInternal()
...
+                        continue;
+                    }
 
-                    continue;
+                    default:
+                        throw connector.Break(new NpgsqlException($"Unknown replication message code '{code}'"));
+                    }
                 }
+            }
+            finally
```
Comment (vonzshik) [inline src/Npgsql/Replication/ReplicationConnection.cs:496]: Just to clarify - if we've failed somewhere in a user code - does it still get called?

File: src/Npgsql/Replication/ReplicationConnection.cs [hunk 2]
```diff
@@ -428,74 +421,100 @@ async Task<TimelineHistoryFile> TimelineHistoryInternal()
...
+                    msg = await connector.ReadMessage(true, CancellationToken.None);
+                    Expect<CopyDataMessage>(msg, connector);
 
-                    // dataLen = msg.Length - (code = 1 + walStart = 8 + walEnd = 8 + serverClock = 8)
-                    var dataLen = messageLength - 25;
-                    columnStream.Init(dataLen, canSeek: false);
-
-                    _cachedXLogDataMessage.Populate(new NpgsqlLogSequenceNumber(startLsn), new NpgsqlLogSequenceNumber(endLsn), sendTime, columnStream);
-                    yield return _cachedXLogDataMessage;
+                    // We received some message so there's no need to forcibly request feedback or time out.
```
Comment (Brar) [inline src/Npgsql/Replication/ReplicationConnection.cs:439]: Since I moved the cancellation timeout to the buffers this comment is somewhat outdated.

```suggestion
                    // We received some message so there's no need to forcibly request feedback
```

File: src/Npgsql/Replication/ReplicationConnection.cs [hunk 3]
```diff
@@ -527,17 +546,13 @@ async Task SendFeedback(bool waitOnSemaphore = false, bool requestReply = false,
...
-                _requestFeedbackTimer.Change(Timeout.InfiniteTimeSpan, Timeout.InfiniteTimeSpan);
-
                 var connector = _npgsqlConnection.Connector!;
                 var buf = connector.WriteBuffer;
 
                 const int len = 39;
 
                 if (buf.WriteSpaceLeft < len)
-                    await connector.Flush(true, cancellationToken);
+                    await connector.Flush(async: true, cancellationToken);
```
Comment (Brar) [inline src/Npgsql/Replication/ReplicationConnection.cs:555]: Was this change intentional. If yes, why?

File: src/Npgsql/Replication/ReplicationConnection.cs [hunk 4]
```diff
@@ -549,16 +564,12 @@ async Task SendFeedback(bool waitOnSemaphore = false, bool requestReply = false,
                 buf.WriteInt64(TimestampHandler.ToPostgresTimestamp(DateTime.Now));
                 buf.WriteByte(requestReply ? (byte)1 : (byte)0);
 
-                await connector.Flush(true, cancellationToken);
+                await connector.Flush(async: true, cancellationToken);
```
Comment (Brar) [inline src/Npgsql/Replication/ReplicationConnection.cs:567]: See above.

File: src/Npgsql/Replication/ReplicationConnection.cs [hunk 5]
```diff
@@ -653,7 +664,9 @@ async Task<object[]> ReadSingleRow(string command, CancellationToken cancellatio
 #if !NETSTANDARD2_0
             await
 #endif
-            using var registration = cancellationToken.CanBeCanceled ? cancellationToken.Register(c => ((ReplicationConnection)c!).Cancel(), this) : default;
+            using var registration = cancellationToken.CanBeCanceled
+                ? cancellationToken.Register(c => ((ReplicationConnection)c!).Connector.CancelRequest(), this)
+                : default;
```
Comment (Brar) [inline src/Npgsql/Replication/ReplicationConnection.cs:669]: See above.
```suggestion
            using var registration = cancellationToken.Register(c => ((NpgsqlConnector)c!).CancelRequest(), connector);
```

File: src/Npgsql/Replication/ReplicationConnection.cs [hunk 6]
```diff
@@ -623,7 +634,7 @@ async Task DropReplicationSlotInternal()
                 await
 #endif
                 using var registration = cancellationToken.CanBeCanceled
-                    ? cancellationToken.Register(c => ((ReplicationConnection)c!).Cancel(), this)
+                    ? cancellationToken.Register(c => ((ReplicationConnection)c!).Connector.CancelRequest(), this)
                     : default;
```
Comment (Brar) [inline src/Npgsql/Replication/ReplicationConnection.cs:638]: You changed this at some other place so let's change it here too.
If I understand it correctly we don't have to check for `cancellationToken.CanBeCanceled` here.
Also let's just pass in the connector we have at hands anyways.
```suggestion
                using var registration = cancellationToken.Register(c => ((NpgsqlConnector)c!).CancelRequest(), connector);
```

File: src/Npgsql/Replication/ReplicationConnection.cs [hunk 7]
```diff
@@ -28,15 +29,14 @@ public abstract class ReplicationConnection : IAsyncDisposable
         static readonly Version FirstVersionWithoutDropSlotDoubleCommandCompleteMessage = new Version(13, 0);
         static readonly Version FirstVersionWithTemporarySlots = new Version(10, 0);
         static readonly NpgsqlLogger Log = NpgsqlLogManager.CreateLogger(nameof(ReplicationConnection));
-        readonly Stopwatch _cancelTimer = new Stopwatch();
```
Comment (Brar) [inline src/Npgsql/Replication/ReplicationConnection.cs:31]: Thanks that was a leftover from the early days.

**Review by Brar — APPROVED** [review]
LGTM.
A few minor comments/suggestions

File: src/Npgsql/Replication/ReplicationConnection.cs (hunk 1 above)
Comment (roji) [reply src/Npgsql/Replication/ReplicationConnection.cs:496]: Whatever happens in user code, this method will still need to exit somehow (i.e. with an exception). So if user code somehow causes the socket to break or something, we'd also end up here - right?

File: src/Npgsql/Replication/ReplicationConnection.cs (hunk 3 above)
Comment (roji) [reply src/Npgsql/Replication/ReplicationConnection.cs:555]: Yeah, intentional but not important - I now tend to use explicit parameter names everywhere where the literal isn't self-explanatory (it's hard to know if you're a new-comer what the `true` means alone).

File: src/Npgsql/Replication/ReplicationConnection.cs (hunk 4 above)
Comment (roji) [reply src/Npgsql/Replication/ReplicationConnection.cs:567]: As above

File: src/Npgsql/Replication/ReplicationConnection.cs (hunk 6 above)
Comment (roji) [reply src/Npgsql/Replication/ReplicationConnection.cs:638]: Makes sense. The CanBeCanceled check really makes sense only in very hot paths (and even there I'm not so sure it matters).

File: src/Npgsql/Replication/ReplicationConnection.cs (hunk 1 above)
Comment (roji) [reply src/Npgsql/Replication/ReplicationConnection.cs:496]: @vonzshik am merging this to move on with other stuff, but let's continue this conversation if necessary - we can fix after this PR if needed.

Comment (vonzshik) [reply src/Npgsql/Replication/ReplicationConnection.cs:499]: Just checked - seems to work fine.

**Review by vonzshik — APPROVED** [review]

**Comment by roji** [comment]
Merged via c781f9de4f26d67c27f269ebb6b82f1c7ea8791e

File: src/Npgsql/Replication/ReplicationConnection.cs (hunk 6 above)
Comment (Brar) [reply src/Npgsql/Replication/ReplicationConnection.cs]: > and even there I'm not so sure it matters

It's pretty close to a no-op for default tokens.
See https://github.com/dotnet/runtime/blob/14926991738435e28dbe5b21d0c7e253b2fff060/src/libraries/System.Private.CoreLib/src/System/Threading/CancellationToken.cs#L301
